crates/legacy/src/builder_state.rs

```diff
@@ -1091,14 +1070,10 @@ impl<TYPES: NodeType> BuilderState<TYPES> {
         while Instant::now() <= timeout_after {
             match self.tx_receiver.try_recv() {
                 Ok(tx) => {
-                    if self.included_txns.contains(&tx.commit)
-                        || self.included_txns_old.contains(&tx.commit)
-                        || self.included_txns_expiring.contains(&tx.commit)
-                        || self.txns_in_queue.contains(&tx.commit)
-                    {
+                    if self.included_txns.contains(&tx.commit) {
                         continue;
                     }
-                    self.txns_in_queue.insert(tx.commit);
+                    self.included_txns.insert(tx.commit);
```

Comment:
IUC we should add a transaction to `included_txns` later when we know it's included (by proposals) rather than here when it's just being collected. It probably won't cause issues in the happy path, but when the collected transactions are the same as the proposed ones, this may cause some inconsistency.

Also, not sure how relevant this is, but _maybe_, this causes `view_txn_count_history` to include duplicate transactions in its `current` and `fresh` fields.

Reply:
Hmm... interesting... I had modeled this after what was being done in the `marketplace` implementation.  Perhaps a similar issue occurs there?  I think without an addition to `included_txns`, you could conceivably end up with duplicate transactions submitted within the same block, couldn't you?

Reply:
> I had modeled this after what was being done in the `marketplace` implementation. Perhaps a similar issue occurs there?

Looks like the `collect_txns` function in `marketplace` is updating `txn_commits_in_queue` (similar to the `txns_in_queue`) rather than `included_txns`, so that should be fine.

> I think without an addition to `included_txns`, you could conceivably end up with duplicate transactions submitted within the same block, couldn't you?

If we update neither `included_txns` nor `txns_in_queue`, I think yes, this may cause duplicate transactions. But in the original implementation, we insert the transaction into `txns_in_queue`, so next time when we get a duplicated transaction, the `contains` check before the `insert` and `push` should fail and we will continue the loop without adding this transaction again.